lib/heap/heap.c

```diff
@@ -459,22 +457,66 @@ void *sys_heap_aligned_realloc(struct sys_heap *heap, void *ptr,
 					  bytes_freed);
 #endif
 
+		return true;
+	}
+
+	return false;
+}
+
+void *sys_heap_realloc(struct sys_heap *heap, void *ptr, size_t bytes)
+{
+	/* special realloc semantics */
+	if (ptr == NULL) {
+		return sys_heap_alloc(heap, bytes);
+	}
+	if (bytes == 0) {
+		sys_heap_free(heap, ptr);
+		return NULL;
+	}
+
+	if (inplace_realloc(heap, ptr, bytes)) {
+		return ptr;
+	}
+
+	/* In-place realloc was not possible: fallback to allocate and copy. */
+	void *ptr2 = sys_heap_alloc(heap, bytes);
+
+	if (ptr2 != NULL) {
+		size_t prev_size = sys_heap_usable_size(heap, ptr);
+
+		memcpy(ptr2, ptr, MIN(prev_size, bytes));
+		sys_heap_free(heap, ptr);
+	}
+	return ptr2;
+}
+
+void *sys_heap_aligned_realloc(struct sys_heap *heap, void *ptr,
+			       size_t align, size_t bytes)
+{
+	/* special realloc semantics */
+	if (ptr == NULL) {
+		return sys_heap_aligned_alloc(heap, align, bytes);
+	}
+	if (bytes == 0) {
+		sys_heap_free(heap, ptr);
+		return NULL;
+	}
+
+	__ASSERT((align & (align - 1)) == 0, "align must be a power of 2");
```

Comment:
Just in case there is another respin
```suggestion
	__ASSERT(IS_POWER_OF_TWO(align));
```

Reply:
Ah, align of 0 would trigger the assert as well and this macro would not work in that case.